### ISSUE TITLE
Mix precision leads to nan loss

### ISSUE THREAD

**Issue by ZhenShuo2021** [tensorflow/tensorflow#56142]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Bug

### Source

pip install

### Tensorflow Version

tf2.8

### Custom Code

Yes

### OS Platform and Distribution

windows

### Mobile device

windows 21h2 (19044, 1706)

### Python version

python 3.9.7

### Bazel version

unkonwn

### GCC/Compiler version

unknown

### CUDA/cuDNN version

11.2

### GPU model and memory

RTX3070/8Gb

### Problem Discription?

Loss is nan when using mix precision API even though use (get_scaled_loss, get_unscaled_gradients) or scale loss manually.

- Model works well in float32.
- Mix precision API with [tensorflow guide](https://www.tensorflow.org/guide/mixed_precision#training_the_model_with_a_custom_training_loop) works well in my computer.
- My model input start from Cond2D -> ... -> Resblock -> ... -> Dense -> SoftMax, it's to solve a classification problem.

### Standalone code to reproduce the issue

```shell
import os
os.environ['TF_CPP_MIN_LOG_LEVEL'] = '3'
import numpy as np
import tensorflow as tf
from utils._history import fig
from utils.network_float16 import MyModel, CosineAnnealingScheduler, CCE, Scale_Gradient
from utils.system import args, Yie, _Plot, _Predict, MyTrain
tf.keras.mixed_precision.set_global_policy('mixed_float16')

test_SNR = np.arange(0, 22, 2) if args.fading_type == 1 else np.arange(10, 32, 2)
a,b = MyTrain(args.K, args.m, args.Nd, args.B, args.p, args.Ka, args.snr, args.tap, args.fading_type)   # training data generation

train_dataset = tf.data.Dataset.from_tensor_slices((a, b)) 
train_dataset = train_dataset.shuffle(buffer_size=2048).batch(args.B)
optimizer = tf.keras.optimizers.Adam(learning_rate=5e-4)
optimizer = tf.keras.mixed_precision.LossScaleOptimizer(optimizer)
loss_metric = tf.keras.metrics.CategoricalCrossentropy()
acc = tf.keras.metrics.Accuracy()

model = MyModel(args)
model = model.Model(is_summary=0, is_plot_model=1)   # compile model


@tf.function
def train_step(x_train, y_train):
    scale=2048
    with tf.GradientTape() as tape:
        y_pred = model(x_train,training = True)
        loss = tf.keras.losses.categorical_crossentropy(y_train, y_pred)
        loss *= scale
    grads = tape.gradient(loss, model.trainable_weights)
    grads = [x*scale for x in grads]
    optimizer.apply_gradients(zip(grads, model.trainable_weights))
    return loss


def decision(y_pred):
    y_pred = np.squeeze(np.argsort(-y_pred)[:, :args.Ka])
    temp = np.zeros([args.B, args.K], dtype='int')
    for j in range(args.Ka):
        temp[np.arange(args.B), y_pred[:, j]] = 1
    return temp

for epoch in range(args.epoch):
    epoch_time = time.time()
    for step, (x_train, y_train) in enumerate(train_dataset):
        loss = train_step(x_train, y_train)
    y_pred = model(x_train)
    y_pred = decision(y_pred)
    acc.update_state(y_pred, y_train)
    print("Epoch %d - %ds - loss: %.4f - acc: %4f" % (epoch+1, time.time()-epoch_time, loss_record[epoch], acc.result().numpy()))

_Plot.Simulation(args, _Predict(args, model, test_SNR), test_SNR, fig, save_txt=1)
##### Loss is nan from first epoch #####



##### Next, I tried two types of train_step, and the 'Revelant log output' is the print output #####
##### First I set the size of x_train to batch size, so the network will only update weight ones #####
##### Compile the model and run the following code #####
model = MyModel(args)
model = model.Model(is_summary=0, is_plot_model=1)   # compile model
# type 1 (follow tensorflow guide)
for scale in range(1,3000,30):   # to check which scale is the best
    with tf.GradientTape() as tape:
        y_pred = model(x_train,training = True)
        loss = tf.keras.losses.categorical_crossentropy(y_train, y_pred)
        print(tf.reduce_mean(loss))    # see if CCE loss works normally
        loss = optimizer.get_scaled_loss(loss)
        print(tf.reduce_mean(loss))   # see if get_scaled_loss works normally
    grads = tape.gradient(loss, model.trainable_weights)
    grads = optimizer.get_unscaled_gradients(grads)
    optimizer.apply_gradients(zip(grads, model.trainable_weights))
# -----------------------------------------------------------------
# type 2 (scale loss manually)
for scale in range(1,3000,30):   # to check which scale is the best
    with tf.GradientTape() as tape:
        y_pred = model(x_train,training = True)
        loss = tf.keras.losses.categorical_crossentropy(y_train, y_pred)
        loss *= scale
        print(tf.reduce_mean(loss))   # see if loss scaled properly
    grads = tape.gradient(loss, model.trainable_weights)
    print(tf.reduce_mean(grads[0]))  #  _ in grads[ _ ] can be arbitrary number (layer)
    grads = [x*scale for x in grads]
    optimizer.apply_gradients(zip(grads, model.trainable_weights))

```

### Relevant log output

```shell
type 1

tf.Tensor(9.516, shape=(), dtype=float16)   # unscaled loss
tf.Tensor(inf, shape=(), dtype=float16)     # scaled loss
tf.Tensor(9.516, shape=(), dtype=float16)   # unscaled loss
tf.Tensor(inf, shape=(), dtype=float16)     # scaled loss
...
... # same for other scales
# -------------------------------------------------------------------------
type 2

tf.Tensor(9.516, shape=(), dtype=float16)    # scaled loss
tf.Tensor(nan, shape=(), dtype=float32 )     # gradients
tf.Tensor(961.0, shape=(), dtype=float16)    # scaled loss
tf.Tensor(nan, shape=(), dtype=float32)      # gradients
tf.Tensor(1913.0, shape=(), dtype=float16)   # scaled loss
tf.Tensor(nan, shape=(), dtype=float32)      # gradients
...
...  # gradients are all nan for all scales

##### All gradients are nan from first step, even w/wo kernel initializer (random normal initializer) #####
```
</details>

**Comment by sushreebarsa** [comment]
@ZhenShuo2021 
In order to expedite the trouble-shooting process, please provide a complete code snippet to reproduce the issue reported here. Thank you!

**Comment by ZhenShuo2021** [comment]
Complete code update, please take a look.
Thank you!

**Comment by sushreebarsa** [comment]
@ZhenShuo2021 Thank you for the update!
Please post this issue on [keras-team/keras repo.](https://github.com/keras-team/keras/issues)
To know more see;
[https://discuss.tensorflow.org/t/keras-project-moved-to-new-repository-in-https-github-com-keras-team-keras/1999](https://discuss.tensorflow.org/t/keras-project-moved-to-new-repository-in-https-github-com-keras-team-keras/1999)
Thank you!
